atlp-rwanda/Phantom-Backend-Elites
# Register, login, logout, roles and permissions, and password reset features all in one

## What does this PR do?
- It adds user registration feature
- It adds user login feature
- Adds logout feature
- Adds Roles and permissions feature
- Finally, it adds password reset functionality 

## Description of Tasks to be completed
1. **User Registration:**
   - [x]  Added a `User Model` for registering users, in the `models` directory.
   - [x]  Added a `Controller` for user model in the `controllers` folder
   - [x]  Added a `userRoutes` file in the routes folder
2. **User login:**
   - [x]  Added login function to the authController  where the login logic is implemented
   - [x]  Click here https://trello.com/c/BRlmfGeF to view more details on this task on trello
3. **User logout:**
   - [x] Added a logout function in the authController to handle the logout logic
   - [x]  Click here https://trello.com/c/cX3FETFS to view more details on this task on trello
4. **Roles and permissions:**
   - [x] Created `Role and Permission` models in the model folder
   - [x] Created controllers for roles and permissions CRUD
   - [x] Added routes for roles and permissions in the routes foleder
   - [x]  Click here https://trello.com/c/0IkWP4tR to view more details on this task on trello
5. **Password Reset:**
   - [x] Added the ResetToken model that will be storing password reset token
   - [x] Created controller to handle the sending password reset links to the user emails and changing the password when the a user clicks on those links

## How should this be manually tested?
- clone the repository, then run `npm i` to install all depandencies.
- start the server with `npm run dev`
- Go to `http://localhost:5000/api/v1/users` to see a list of all users. you can add `/id` to grab a specific user, delete or update
- Go to `http://localhost:5000/api/v1/roles` to see a list of all roles. you can add `/id` to grab a specific role, delete or update
- Go to `http://localhost:5000/api/v1/permissions` to see a list of all permissions. you can add `/id` to grab a specific permission, delete or update
- Go to `http://localhost:5000/api/v1/auth/login` to log into the system
- Go to `http://localhost:5000/api/v1/auth/logout` to log out the system
- Go to `http://localhost:5000/api/v1/reset-password/link` to send the password reset link on the `email` supplied in the body of the request
- Go to `http://localhost:5000/api/v1/reset-password/new-password` to change your password. It requires to supply the token sent to the email, password and confirmPassword in the body of the request.

- [x] `Note:` All of the above endpoints are secured by jwt authorization token. A token is generated at user login, and it should supplied in the headers of the request to be authorized.

## What clients did you use to test this?
- [x] Postman
